src/enclave/gencreds.c

```diff
@@ -0,0 +1,415 @@
+        goto done;
+    }
+
+    if ((ret = oe_generate_attestation_certificate(
+             (unsigned char*)common_name,
+             private_key,
+             private_key_size,
+             public_key,
+             public_key_size,
+             &cert,
+             &cert_size)) != OE_OK)
+    {
+        result = ret;
+        goto done;
+    }
+
+    *private_key_out = private_key;
+    *private_key_size_out = private_key_size;
+    private_key = NULL;
+
+    *cert_out = cert;
+    *cert_size_out = cert_size;
+    cert = NULL;
+
+    result = OE_OK;
+
+done:
+
+    if (private_key)
+        oe_free_key(private_key, private_key_size, NULL, 0);
+
+    if (public_key)
+        oe_free_key(public_key, public_key_size, NULL, 0);
+
+    if (cert)
+        oe_free_attestation_certificate(cert);
+
+    return result;
+}
+
+static int _generate_attested_credentials(
+    const char* common_name,
+    uint8_t** cert_out,
+    size_t* cert_size_out,
+    uint8_t** private_key_out,
+    size_t* private_key_size_out)
+{
+    int ret = -1;
+    uint8_t* cert = NULL;
+    size_t cert_size;
+    uint8_t* private_key = NULL;
+    size_t private_key_size;
+
+    if (cert_out)
+        *cert_out = NULL;
+
+    if (cert_size_out)
+        *cert_size_out = 0;
+
+    if (private_key_out)
+        *private_key_out = NULL;
+
+    if (private_key_size_out)
+        *private_key_size_out = 0;
+
+    if (!cert_out || !cert_size_out || !private_key_out ||
+        !private_key_size_out)
+    {
+        goto done;
+    }
+
+    /* Generate the attested certificate and private key */
+    if (_generate_cert_and_private_key(
+            common_name, &cert, &cert_size, &private_key, &private_key_size) !=
+        OE_OK)
+    {
+        goto done;
+    }
+
+    /* Verify that the certificate can be parsed as DER */
+    {
+        mbedtls_x509_crt crt;
+        mbedtls_x509_crt_init(&crt);
+
+        if (mbedtls_x509_crt_parse_der(&crt, cert, cert_size) != 0)
+        {
+            mbedtls_x509_crt_free(&crt);
+            goto done;
+        }
+
+        mbedtls_x509_crt_free(&crt);
+    }
+
+    /* Verify that the private key can be parsed as PEM */
+    {
+        mbedtls_pk_context pk;
+        mbedtls_pk_init(&pk);
+
+        if (mbedtls_pk_parse_key(&pk, private_key, private_key_size, NULL, 0) !=
+            0)
+        {
+            mbedtls_pk_free(&pk);
+            goto done;
+        }
+
+        mbedtls_pk_free(&pk);
+    }
+
+    *cert_out = cert;
+    cert = NULL;
+    *cert_size_out = cert_size;
+    *private_key_out = private_key;
+    private_key = NULL;
+    *private_key_size_out = private_key_size;
+
+    ret = 0;
+
+done:
+
+    if (cert)
+        oe_free_key(cert, cert_size, NULL, 0);
+
+    if (private_key)
+        oe_free_key(private_key, private_key_size, NULL, 0);
+
+    if (cert)
+        oe_free_attestation_certificate(cert);
+
+    return ret;
+}
+
+
+static int _write_file(const char* path, const void* data, size_t size)
```

Comment:
Both this (_write_file) and _load_file are pretty general utility functions. Do we have a more common place for them than gencreds.c?